### ISSUE TITLE
worker->submitTo() on new Pool() and ->resize()

### ISSUE THREAD

**Issue by cottton** [krakjoe/pthreads#462]
Im testing with the Pool class and found out that the workers are not started on new Pool() or ->resize().
This makes it difficult or impossible to use ->submitTo() on a "fresh" Pool.

I expected new Pool(2) creates 2 workers at offsets 0 and 1.
But $this->workers is NULL.
Only if submitted work to the Pool and all workers need to be used they get created.

let me show you what i mean:
```php
class Task extends Collectable {}
class P extends Pool
{
    public function test()
    {
        echo PHP_EOL . __METHOD__ . PHP_EOL;
        echo "size: "; var_dump($this->size);
        echo "workers: "; var_dump($this->workers);
        echo "work: "; var_dump($this->work);
    }
}
$P = new P(2);
$P->submit(new Task());
$P->test();
echo PHP_EOL . '->resize(4);' . PHP_EOL;
$P->resize(4);
$P->test();

echo "\n---------------\n\ncall ->submitTo(2, new Task())\n";
$P->submitTo(2, new Task()); // fatal error "The selected worker (2) does not exist"
$P->test();
/* output
P::test
size: int(2)
workers: array(1) {
  [0]=>
  object(Worker)#4 (0) {
  }
}
work: array(1) {
  [0]=>
  object(Task)#3 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#4 (0) {
    }
  }
}

->resize(4);

P::test
size: int(4)
workers: array(1) {
  [0]=>
  object(Worker)#4 (0) {
  }
}
work: array(1) {
  [0]=>
  object(Task)#3 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#4 (0) {
    }
  }
}

---------------

call ->submitTo(2, new Task())

Fatal error: Uncaught exception 'Exception' with message 'The selected worker (2) does not exist' in ...
*/
```
with a dirty trick i could get it working as expected:
```php
class Task extends Collectable {}
class Dummy extends Collectable {}
class P extends Pool
{
    public function test()
    {
        echo PHP_EOL . __METHOD__ . PHP_EOL;
        echo "size: "; var_dump($this->size);
        echo "workers: "; var_dump($this->workers);
        echo "work: "; var_dump($this->work);
    }
    public function fixedResisze($i)
    {
        $this->resize((int)$i);
        while(count($this->workers) < $this->size){
            $this->submit(new Dummy());
            $this->collect(
                function($Collectable){
                    if(is_a($Collectable, 'Dummy')){
                        $Collectable->setGarbage();
                        return $Collectable->isGarbage();
                    }
                }
            );
        }
    }
}
$P = new P(2);
$P->submit(new Task());
$P->test();
echo PHP_EOL . '->fixedResisze(4);' . PHP_EOL;
$P->fixedResisze(4);
$P->test();

echo "\n---------------\n\ncall ->submitTo(2, new Task())\n";
$P->submitTo(2, new Task());
$P->test();
/* output
P::test
size: int(2)
workers: array(1) {
  [0]=>
  object(Worker)#4 (0) {
  }
}
work: array(1) {
  [0]=>
  object(Task)#3 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#4 (0) {
    }
  }
}

->fixedResisze(4);

P::test
size: int(4)
workers: array(4) {
  [0]=>
  object(Worker)#4 (0) {
  }
  [1]=>
  object(Worker)#6 (0) {
  }
  [2]=>
  object(Worker)#8 (0) {
  }
  [3]=>
  object(Worker)#10 (0) {
  }
}
work: array(1) {
  [0]=>
  object(Task)#3 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#4 (0) {
    }
  }
}

---------------

call ->submitTo(2, new Task())

P::test
size: int(4)
workers: array(4) {
  [0]=>
  object(Worker)#4 (0) {
  }
  [1]=>
  object(Worker)#6 (0) {
  }
  [2]=>
  object(Worker)#8 (0) {
  }
  [3]=>
  object(Worker)#10 (0) {
  }
}
work: array(2) {
  [0]=>
  object(Task)#3 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#4 (0) {
    }
  }
  [4]=>
  object(Task)#11 (2) {
    ["garbage"]=>
    bool(false)
    ["worker"]=>
    object(Worker)#8 (0) {
    }
  }
}
*/
```
would be great if 
new Pool(n) 
and
->resize(n)
creates n workers, or at least the offsets to be able to use ->submitTo().

**Comment by JakeFalcor** [comment]
It's actually a feature - obviously there is no need to construct Workers since there we're no tasks submitted?

This applies to #463 as well.

**Comment by cottton** [comment]
The problem is if you try to submitTo() a worker you will fail if __not__ the full amount of workers was ever needed.
Im currently writing a Pool that handles ident based workers for sockets.
Means: 
- many sockets (clients) to be served
- all have to get data in correct order (cannot just Pool->submit() because workers would send packets in random order and client-scripts are "stupid" =)
- each worker gets n clients to serve

Example:
got packet for socket
asking 1st worker if is serving that socket like 
```php
foreach($this->workers as $i => $worker){
}
```
and it fails because `$this->workers` === null.
Also not working:
- got 3 worker
- all worker busy (max idents reached)
- gona resize Pool +1 and push that ident to the new worker:
```php
// ...
$this->resize($this->size+1);
$my_new_worker = $this->workers[$this->last]; // fail!
```
(See parameter $last bug posted at https://github.com/krakjoe/pthreads/issues/463)
So this will fail because: `$this->last` is "not updated", and the new worker is not created.
The selected worker in example above would be the last worker we checked -- which was actually busy.
If para $last would have been updated we would get a problem (and thats my actual problem)
with 
```php
// we need a new worker 
$this->resize($this->size+1); 
// get offset of new worker
$offset = $this->last; // === second last
$offset = $this->size-1; // this offset doesnt exist 
$offset = count($this->workers)-1; // === second last
// no matter what ... offset is wrong
$my_new_worker = $this->workers[$offset]; // fails
$this->submitTo($offset, $work); // fails
```
Solution(?):
On ->resize() create offset at $this->workers and on ->submitTo() check if worker was already started?
